File: pynurbs/geometry/methods/intersect_triangle.py
from __future__ import division

from numpy import dot, cross, zeros, float64
from numpy.linalg import norm

from pynurbs.config import Settings


def intersect_triangle_ray(v0, v1, v2, origin, ray, test_cull=False, eps=None):
    """
    Compute the intersection of a triangle with a ray.

    :param ndarray v0: Vertex 0 of triangle.
    :param ndarray v1: Vertex 1 of triangle.
    :param ndarray v2: Vertex 2 of triangle.
    :param ndarray origin: Ray origin.
    :param ndarray ray: Ray direction (unit vector).
    :param bool test_cull: Option to cull back facing triangles.
    :param float eps: Tolerance for ray-triangle intersection. If *None* is
        provided, then the *eps* value from the settings class is used.

    :return: *None* if the ray does not intersect the triangle. If an
        intersection is found, then the parameter *t* on the ray and and
        parameters (u,v) on the triangle are returned (t, u, v).
    :rtype: tuple

    *Reference:* Moller, T., Trumbore, B., "Fast Minimum Storage Ray Triangle
    Intersection."
    """
    if eps is None:
        eps = Settings.atol

    e1 = v1 - v0
    e2 = v2 - v0
    pvec = cross(ray, e2)
    det = dot(e1, pvec)
    if test_cull:
        if det <= eps:
            return None
        tvec = origin - v0
        u = dot(tvec, pvec)
        if u <= -eps or u >= det + eps:
            return None
        qvec = cross(tvec, e1)
        v = dot(ray, qvec)
        if v <= -eps or u + v >= det:
            return None
        t = dot(e2, qvec)
        inv_det = 1. / det
        t *= inv_det
        u *= inv_det
        v *= inv_det
    else:
        if abs(det) < eps:
            return None
        inv_det = 1. / det
        tvec = origin - v0
        u = dot(tvec, pvec) * inv_det
        if u <= -eps or u >= 1. + eps:
            return None
        qvec = cross(tvec, e1)
        v = dot(ray, qvec) * inv_det
        if v <= -eps or u + v >= 1. + eps:
            return None
        t = dot(e2, qvec) * inv_det
    return t, u, v


def intersect_triangles(t1, t2, tol=None, eps=None):
    """
    Find the intersection segment of two triangles.

    :param ndarray t1: Array containing vertices of triangle 1 (3 x 3).
    :param ndarray t2: Array containing vertices of triangle 2 (3 x 3).
    :param float tol: Optional tolerance to use for length of intersection
        segment. If none is provided then the default *gtol* will be used.
    :param float eps: Tolerance for ray-triangle intersection. If *None* is
        provided, then the *eps* value from the settings class is used.

    :return: The number of intersection points and the intersection points that
        form the intersection segment between two triangles (if any) as a
        (2 x 3) array (ni, pi).
    :rtype: tuple
    """
    if tol is None:
        tol = Settings.gtol
    if eps is None:
        eps = Settings.atol

    # Find the normal vectors of the planes of the triangles.
    vn1 = cross(t1[1] - t1[0], t1[2] - t1[0])
    vn2 = cross(t2[1] - t2[0], t2[2] - t2[0])

    # TODO Test for parallel triangles.
    # if abs(dot(vn1, vn2)) - norm(vn1) * norm(vn2) <= Settings.atol:
    #     return 0, []

    # Test the two triangles for possible intersect by checking the signed
    # distance of their vertices to the other's plane.
    test = False
    d0 = dot(vn1, t2[0] - t1[0])
    for i in range(1, 3):
        di = dot(vn1, t2[i] - t1[0])
        if d0 * di <= 0.:
            test = True
            break
    if not test:
        return 0, []

    test = False
    d0 = dot(vn2, t1[0] - t2[0])
    for i in range(1, 3):
        di = dot(vn2, t1[i] - t2[0])
        if d0 * di <= 0.:
            test = True
            break
    if not test:
        return 0, []

    # Intersect edges of triangle 1 with triangle 2.
    verts = zeros((3, 3), dtype=float64)
    edges = zeros((3, 3), dtype=float64)
    edges[0] = t1[1] - t1[0]
    verts[0] = t1[0]
    edges[1] = t1[2] - t1[0]
    verts[1] = t1[0]
    edges[2] = t1[2] - t1[1]
    verts[2] = t1[1]
    pnts = zeros((2, 3), dtype=float64)
    npts = 0
    for i in range(3):
        tuv = intersect_triangle_ray(t2[0], t2[1], t2[2], verts[i], edges[i])
        if tuv is None or tuv[0] < -eps or tuv[0] > 1. + eps:
            continue
        t, u, v = tuv
        # pi = (1. - u - v) * t2[0] + u * t2[1] + v * t2[2]
        pi = verts[i] + t * edges[i]
        pnts[npts] = pi
        npts += 1
        if npts == 2:
            # Keep going if there are vertices still to check.
            di = norm(pnts[1] - pnts[0])
            if di <= tol and i < 2:
                npts -= 1
            elif di > tol:
                return npts, pnts
            elif di <= tol and i == 2:
                npts -= 1

    # Intersect edges of triangle 2 with triangle 1.
    edges[0] = t2[1] - t2[0]
    verts[0] = t2[0]
    edges[1] = t2[2] - t2[0]
    verts[1] = t2[0]
    edges[2] = t2[2] - t2[1]
    verts[2] = t2[1]
    for i in range(3):
        tuv = intersect_triangle_ray(t1[0], t1[1], t1[2], verts[i], edges[i])
        if tuv is None or tuv[0] < -eps or tuv[0] > 1. + eps:
            continue
        t, u, v = tuv
        # pi = (1. - u - v) * t1[0] + u * t1[1] + v * t1[2]
        pi = verts[i] + t * edges[i]
        pnts[npts] = pi
        npts += 1
        if npts == 2:
            # Keep going if there are vertices still to check.
            di = norm(pnts[1] - pnts[0])
            if di <= tol and i < 2:
                npts -= 1
            elif di > tol:
                return npts, pnts
            elif di <= tol and i == 2:
                npts -= 1

    return 0, []


def intersect_triangle_plane(t, p0, pnorm, tol=None):
    """
    Find the intersection segment of a plane and a triangle.

    :param ndarray t: Array containing vertices of triangle (3 x 3).
    :param ndarray p0: Origin of plane.
    :param ndarray pnorm: Normal vector of plane.
    :param float tol: Optional tolerance to use for length of intersection
        segment. If none is provided then the default *gtol* will be used.

    :return: Intersection segment between triangle and plane (if any) as a
        (2 x 3) array.
    :rtype: ndarray
    """
    if tol is None:
        tol = Settings.gtol

    pnorm /= norm(pnorm)

    # Check for parallel triangle and plane. Return no intersection if
    # they're parallel.
    tri_norm = cross(t[1] - t[0], t[2] - t[0])
    tri_norm /= norm(tri_norm)
    if 1. - abs(dot(pnorm, tri_norm)) <= 1.0e-12:
        return 0, zeros((2, 3), dtype=float64)

    # Test for possible intersection by checking the signed distance of the
    # vertices to the plane.
    test = False
    d0 = dot(pnorm, t[0] - p0)
    for i in range(1, 3):
        di = dot(pnorm, t[i] - p0)
        if d0 * di <= 0.:
            test = True
            break
    if not test:
        return 0, zeros((2, 3), dtype=float64)

    # Intersect the edges of the triangle with the plane.
    verts = zeros((3, 3), dtype=float64)
    edges = zeros((3, 3), dtype=float64)
    edges[0] = t[1] - t[0]
    verts[0] = t[0]
    edges[1] = t[2] - t[0]
    verts[1] = t[0]
    edges[2] = t[2] - t[1]
    verts[2] = t[1]
    pnts = zeros((2, 3), dtype=float64)
    npts = 0
    for i in range(3):
        d0 = dot(pnorm, verts[i] - p0)
        d1 = dot(pnorm, verts[i] + edges[i] - p0)
        if d0 * d1 > 0.:
            continue
        if abs(d0) <= tol and abs(d1) <= tol:
            pnts[0] = verts[i]
            pnts[1] = verts[i] + edges[i]
            return 2, pnts
        denom = dot(pnorm, edges[i])
        ui = dot(pnorm, p0 - verts[i]) / denom
        if ui < 0.:
            ui = 0.
        if ui > 1.:
            ui = 1.
        pi = verts[i] + ui * edges[i]
        pnts[npts] = pi
        npts += 1
        if npts == 2:
            # Keep going if there are vertices still to check.
            di = norm(pnts[1] - pnts[0])
            if di <= tol and i < 2:
                npts -= 1
            elif di <= tol and i == 2:
                return 0, zeros((2, 3), dtype=float64)
            else:
                return npts, pnts
    return 0, zeros((2, 3), dtype=float64)

File: src/geometry/intersect_triangle.f90
module intersect_triangle
use math
implicit none

private
public :: intersect_triangle_ray, intersect_triangle_plane, intersect_triangles

contains

subroutine intersect_triangle_ray(v0, v1, v2, origin, ray, eps, &
                                  test_cull, t, u, v, success)
    !> Compute the intersection of a triangle with a ray.
    !> v0 - Vertex 0 of triangle.
    !> v1 - Vertex 1 of triangle.
    !> v2 - Vertex 2 of triangle.
    !> origin - Ray origin.
    !> ray - Ray direction (unit vector).
    !> eps - Nondimensional tolerance used when a ray intersects near the
    !> edges of a triangle.
    !> test_cull - Option to cull back facing triangles.
    !> t - Intersection parameter on ray.
    !> u - Barycentric parameter on triangle.
    !> v - Barycentric parameter on triangle.
    !> success - Logical to specify if an intersection was found at all.
    
    !f2py intent(in) v0, v1, v2, origin, ray, eps, test_cull
    !f2py intent(out) t, u, v, success
    
    ! Input
    logical, intent(in) :: test_cull
    double precision, intent(in) :: eps
    double precision, intent(in) :: v0(0:2), v1(0:2), v2(0:2)
    double precision, intent(in) :: origin(0:2), ray(0:2)
    
    ! Output
    logical, intent(out) :: success
    double precision, intent(out) :: t, u, v
    
    ! Working
    double precision :: det, inv_det
    double precision :: e1(0:2), e2(0:2), pvec(0:2), tvec(0:2), qvec(0:2)
                        
    t = 0.0d0
    u = 0.0d0
    v = 0.0d0
    success = .true.
    
    e1 = v1 - v0
    e2 = v2 - v0
    pvec = cross(ray, e2)
    det = dot_product(e1, pvec)
    if (test_cull) then
        if (det .le. eps) then
            success = .false.
            return
        end if
        tvec = origin - v0
        u = dot_product(tvec, pvec)
        if ((u .le. -eps) .or. (u .ge. det + eps)) then
            success = .false.
            return
        end if
        qvec = cross(tvec, e1)
        v = dot_product(ray, qvec)
        if ((v .le. -eps) .or. (u + v .ge. det + eps)) then
            success = .false.
            return
        end if
        t = dot_product(e2, qvec)
        inv_det = 1.0d0 / det
        t = t * inv_det
        u = u * inv_det
        v = v * inv_det
    else
        if (abs(det) .lt. eps) then
            success = .false.
            return
        end if
        inv_det = 1.0d0 / det
        tvec = origin - v0
        u = dot_product(tvec, pvec) * inv_det
        if ((u .le. -eps) .or. (u .ge. 1.0d0 + eps)) then
            success = .false.
            return
        end if
        qvec = cross(tvec, e1)
        v = dot_product(ray, qvec) * inv_det
        if ((v .le. -eps) .or. (u + v .ge. 1.0d0 + eps)) then
            success = .false.
            return
        end if
        t = dot_product(e2, qvec) * inv_det
    end if
    
end subroutine intersect_triangle_ray

subroutine intersect_triangles(t1, t2, tol, eps, npts, pnts)
    !> Find the intersection segment between two triangles.
    !> t1 - Vertices of triangle 1.
    !> t2 - Vertices of triangle 2.
    !> tol - Geometric tolerance.
    !> eps - Nondimensional tolerance.
    !> npts - Number of intersection points. This should be two
    !> if a valid intersection segment was found, zero otherwise.
    !> pnts - Intersection points.
    
    !f2py intent(in) t1, t2, tol, eps
    !f2py intent(out) npts, pnts
    
    ! Input
    double precision, intent(in) :: tol, eps
    double precision, intent(in) :: t1(0:2, 0:2), t2(0:2, 0:2)
    
    ! Ouput
    integer, intent(out) :: npts
    double precision, intent(out) :: pnts(0:1, 0:2)
    
    ! Working
    logical :: test
    integer :: i
    double precision :: d0, di, t, u, v
    double precision :: vn1(0:2), vn2(0:2), verts(0:2, 0:2)
    double precision :: edges(0:2, 0:2), e1(0:2), e2(0:2)
    double precision :: v0(0:2), v1(0:2), v2(0:2)
    double precision ::  p0(0:2), ray(0:2), p12(0:2)
                        
    npts = 0
    pnts(:, :) = 0.0d0
    
    e1 = t1(1, :) - t1(0, :)
    e2 = t1(2, :) - t1(0, :)
    vn1 = cross(e1, e2)
    e1 = t2(1, :) - t2(0, :)
    e2 = t2(2, :) - t2(0, :)
    vn2 = cross(e1, e2)
    
    test = .false.
    d0 = dot_product(vn1, t2(0, :) - t1(0, :))
    do i = 1, 2
        di = dot_product(vn1, t2(i, :) - t1(0, :))
        if (d0 * di .le. 0.0d0) then
            test = .true.
            exit
        end if
    end do
    if (test .eqv. .false.) then
        return
    end if
    
    test = .false.
    d0 = dot_product(vn2, t1(0, :) - t2(0, :))
    do i = 1, 2
        di = dot_product(vn2, t1(i, :) - t2(0, :))
        if (d0 * di .le. 0.0d0) then
            test = .true.
            exit
        end if
    end do
    if (test .eqv. .false.) then
        return
    end if
    
    ! Intersect edges of triangle 1 with triangle 2.
    verts(:, :) = 0.0d0
    edges(:, :) = 0.0d0
    edges(0, :) = t1(1, :) - t1(0, :)
    verts(0, :) = t1(0, :)
    edges(1, :) = t1(2, :) - t1(0, :)
    verts(1, :) = t1(0, :)
    edges(2, :) = t1(2, :) - t1(1, :)
    verts(2, :) = t1(1, :)
    v0 = t2(0, :)
    v1 = t2(1, :)
    v2 = t2(2, :)
    do i = 0, 2
        p0 = verts(i, :)
        ray = edges(i, :)
        call intersect_triangle_ray(v0, v1, v2, p0, ray, eps, .false., &
                                    t, u, v, test)
        if ((test .eqv. .true.) .and. &
            (t .ge. -eps) .and. (t .le. 1.0d0 + eps)) then
            pnts(npts, :) = p0 + t * ray
            npts = npts + 1
            if (npts .eq. 2) then
                p12 = pnts(1, :) - pnts(0, :)
                di = norm(p12)
                if ((di .le. tol) .and. (i .lt. 2)) then
                    npts = npts - 1
                elseif (di .gt. tol) then
                    return
                elseif ((di .le. tol) .and. (i .eq. 2)) then
                    npts = npts - 1
                end if
            end if
        end if
    end do
    
    ! Intersect edges of triangle 2 with triangle 1.
    edges(0, :) = t2(1, :) - t2(0, :)
    verts(0, :) = t2(0, :)
    edges(1, :) = t2(2, :) - t2(0, :)
    verts(1, :) = t2(0, :)
    edges(2, :) = t2(2, :) - t2(1, :)
    verts(2, :) = t2(1, :)
    v0 = t1(0, :)
    v1 = t1(1, :)
    v2 = t1(2, :)
    do i = 0, 2
        p0 = verts(i, :)
        ray = edges(i, :)
        call intersect_triangle_ray(v0, v1, v2, p0, ray, eps, &
                                    .false., t, u, v, test)
        if ((test .eqv. .true.) .and. &
            (t .ge. -eps) .and. (t .le. 1.0d0 + eps)) then
            pnts(npts, :) = p0 + t * ray
            npts = npts + 1
            if (npts .eq. 2) then
                p12 = pnts(1, :) - pnts(0, :)
                di = norm(p12)
                if ((di .le. tol) .and. (i .lt. 2)) then
                    npts = npts - 1
                elseif (di .gt. tol) then
                    return
                elseif ((di .le. tol) .and. (i .eq. 2)) then
                    npts = npts - 1
                end if
            end if
        end if
    end do
    ! No intersection found.
    npts = 0
    pnts(:, :) = 0.0d0
end subroutine intersect_triangles

subroutine intersect_triangle_plane(t, p0, pnorm, gtol, itol, ni, pi)
    !> Find the intersection segment of a triangle and a plane.
    !> t - Vertices of triangle.
    !> p0 - Origin of plane.
    !> pnorm - Unit normal vector of plane.
    !> gtol - Geometric tolerance.
    !> itol - Intersection tolerance for checking minimum length of
    !> intersection segment.
    !> ni - Number of intersection points. This should be two if a valid
    !> intersection segment was found, zero otherwise.
    !> pi - Intersection points.
    
    !f2py intent(in) t, p0, pnorm, gtol, itol
    !f2py intent(out) ni, pi
    
    !Input
    double precision, intent(in) :: gtol, itol
    double precision, intent(in) :: t(0:2, 0:2), p0(0:2), pnorm(0:2)
                                    
    ! Output
    integer, intent(out) :: ni
    double precision, intent(out) :: pi(0:1, 0:2)
    
    ! Working
    logical :: test
    integer :: i
    double precision :: d0, d1, di, denom, ui, mag
    double precision :: verts(0:2, 0:2), edges(0:2, 0:2), p(0:2), v(0:2)
    double precision :: v1(0:2), v2(0:2), tri_norm(0:2)
    
    ni = 0
    pi(:, :) = 0.0d0
    
    ! Check for parallel triangle and plane. Return no intersection if
    ! they're parallel.
    v1 = t(1, :) - t(0, :)
    v2 = t(2, :) - t(0, :)
    tri_norm = cross(v1, v2)
    mag = norm(tri_norm)
    tri_norm = tri_norm / mag
    if (1.0d0 - dabs(dot_product(pnorm, tri_norm)) .le. 1.0d-12) then
        return
    end if
    
    ! Test for possible intersection.
    test = .false.
    d0 = dot_product(pnorm, t(0, :) - p0)
    do i = 1, 2
        di = dot_product(pnorm, t(i, :) - p0)
        if (d0 * di .le. 0.0d0) then
            test = .true.
            exit
        end if
    end do
    if (test .eqv. .false.) then
        return
    end if
    
    ! Build array of vertices and edges.
    edges(0, :) = t(1, :) - t(0, :)
    verts(0, :) = t(0, :)
    edges(1, :) = t(2, :) - t(0, :)
    verts(1, :) = t(0, :)
    edges(2, :) = t(2, :) - t(1, :)
    verts(2, :) = t(1, :)
    
    ! Intersect edges with plane.
    do i = 0, 2
        d0 = dot_product(pnorm, verts(i, :) - p0)
        d1 = dot_product(pnorm, verts(i, :) + edges(i, :) - p0)
        if ((dabs(d0) .le. gtol) .and. (dabs(d1) .le. gtol)) then
            pi(0, :) = verts(i, :)
            pi(1, :) = verts(i, :) + edges(i, :)
            ni = 2
            return
        end if
        if (d0 * d1 .le. 0.0d0) then
            ! Find intersection
            denom = dot_product(pnorm, edges(i, :))
            ui = dot_product(pnorm, p0 - verts(i, :)) / denom
            if (ui .lt. 0.0d0) then
                ui = 0.0d0
            end if
            if (ui .gt. 1.0d0) then
                ui = 1.0d0
            end if
            p = verts(i, :) + ui * edges(i, :)
            pi(ni, :) = p
            ni = ni + 1
            if (ni .eq. 2) then
                v = pi(1, :) - pi(0, :)
                di = norm(v)
                if ((di .le. itol) .and. (i .lt. 2)) then
                    ni = ni - 1
                elseif ((di .le. itol) .and. (i .eq. 2)) then
                    ni = 0
                    pi(:, :) = 0.0d0
                    return
                else
                    return
                end if
            end if
        end if
    end do
    ni = 0
end subroutine intersect_triangle_plane

end module intersect_triangle